servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/MulticastPublisher.java

```diff
@@ -15,355 +15,560 @@
-            final int subscriberCount = this.subscriberCount;
-            if (subscriberCount == subscribers.length() || subscriberCount < 0) {
+        void addSubscriber(Subscriber<? super T> subscriber, SignalOffloader signalOffloader,
+                           AsyncContextMap contextMap, AsyncContextProvider contextProvider) {
+            final int sCount = subscribeCountUpdater.incrementAndGet(this);
+            if (exactlyMinSubscribers && sCount > minSubscribers) {
                 deliverOnSubscribeAndOnError(subscriber, signalOffloader, contextMap, contextProvider,
-                        new RejectedSubscribeException("Only " + subscribers.length() + " subscribers are allowed!"));
-                break;
+                        new RejectedSubscribeException("Only " + minSubscribers + " subscribers are allowed!"));
+                return;
             }
-
-            if (subscriberCountUpdater.compareAndSet(this, subscriberCount, subscriberCount + 1)) {
-                MulticastSubscriber<T> multicastSubscriber = new MulticastSubscriber<>(this, subscriber,
-                        subscriberCount);
-                subscribers.set(subscriberCount, multicastSubscriber);
-                multicastSubscriber.onSubscribe(subscription);
-                if (subscriberCount == subscribers.length() - 1) {
-                    // This operator has special behavior where it chooses to use the AsyncContext and signal offloader
-                    // from the last subscribe operation.
-                    original.delegateSubscribe(this, signalOffloader, contextMap, contextProvider);
+            MulticastFixedSubscriber<T> multiSubscriber =
+                    new MulticastFixedSubscriber<>(this, subscriber, contextMap, contextProvider, sCount);
+            for (;;) {
+                final Subscriber<? super T>[] currSubs = subscribers;
+                if (currSubs.length != 0 && currSubs[0] instanceof TerminalSubscriber) {
+                    ((TerminalSubscriber<?>) currSubs[0]).terminate(subscriber);
+                    break;
+                } else {
+                    @SuppressWarnings("unchecked")
+                    Subscriber<? super T>[] newSubs = (Subscriber<? super T>[])
+                            Array.newInstance(Subscriber.class, currSubs.length + 1);
+                    System.arraycopy(currSubs, 0, newSubs, 0, currSubs.length);
+                    newSubs[currSubs.length] = multiSubscriber;
+                    if (newSubscribersUpdater.compareAndSet(this, currSubs, newSubs)) {
+                        addSubscriber(multiSubscriber);
+                        if (sCount == minSubscribers) {
+                            // This operator has special behavior where it chooses to use the AsyncContext and signal
+                            // offloader from the last subscribe operation.
+                            original.delegateSubscribe(this, signalOffloader, contextMap, contextProvider);
```

Comment:
I kinda like more the 'connect' approach from reactivex definition. Gives more dynamic options to the user to choose when to subscribe upstream. Did you consider a Completable in the API?

Reply:
discussed offline. ServiceTalk sources are all lazy so returning a `ConnectablePublisher` won't be very useful as the user doesn't have trigger points to invoke these methods (they are driven by events like subscribing, etc.). We can always add method overloads in the future to provide more flexibility and expose internal state (e.g. number subscribers, cancelations, etc.) as required, but for this PR I'll just stick to the basics.